EQL: Sequences will now support nano-timestamps

bpintea opened the pull request:
With this change nanosecond-resolution timestamps can be used to drive
the EQL sequence query.

Closes #68812.

bpintea committed:
Sequences will now support nano-timestamps

With this change nanosecond-resolution timestamps can be used to drive
the EQL sequence query.

astefan reviewed (commented):
Left few comments.

bpintea committed:
Merge branch 'master' into feat/eql_sequence_nanos

bpintea committed:
Adress review comments

- replace BigDecimal with Instant.
- undo implicitTiebreaker changes.

astefan commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:
```diff
@@ -175,14 +178,28 @@ private static void loadData(RestHighLevelClient client, String indexName, boole
+        Object object = entry.get("timestamp");
+        assertThat(object, instanceOf(Long.class));
+        // interpret the value as nanos since the unix epoch
+        String timestamp = entry.get("timestamp").toString();
```
I meant
```suggestion
        String timestamp = object.toString();
```

bpintea replied:
Ah, you're right, thanks (_inside_ vs _assigned_ threw me off).
I'll fix this with any next round of changes.

Luegg commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -339,4 +340,27 @@ public String toString() {
                 stageToKeys);
     }
+
+    private static long timestampDiffNanos(Object ts1, Object ts2) {
```
Maybe it would be good to have some dedicated unit tests for this method. Does it support diffing over the fully supported range of timestamps? Also, I can imagine the method is vulnerable to long overflow issues.

bpintea replied:
This has been now refactored and mostly covered by OrdinalTests.java.

Luegg commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -339,4 +340,27 @@ public String toString() {
+            } else {
+                diff = i1.minusMillis(((Number) ts2).longValue());
+            }
+            timestampDiffNanos = diff.getEpochSecond() * 1_000_000_000L + diff.getNano();
```
```suggestion
            return diff.getEpochSecond() * 1_000_000_000L + diff.getNano();
```

and the same for the other branches.

bpintea replied:
Refactored.

Luegg commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/search/OrdinalTests.java:
```diff
@@ -83,51 +85,74 @@ public void testCompareToSameTsSameTieNullDifferentImplicitTb() {
+
+    static Object randomTimestamp() {
+        if (randomBoolean()) {
+            return randomLongBetween(-3155701416721920000L, 3155688986440319900L);
```
where do these bounds come from? Will timestamps never be outside of these ranges?

bpintea replied:
Those came from Instant, that has some (private) ranges for seconds. But this has been now improved.

Luegg commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java:
```diff
@@ -9,26 +9,33 @@
     private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(Ordinal.class);
 
-    private final long timestamp;
+    private final Object timestamp;
```
I'm probably missing quite some context, but why not always using instant for holding the ts?

bpintea replied:
Mostly for performance reasons, since timestamp handling is a core sequences operation and presumably most queries run on millis (as it had so far), i.e. longs.

astefan commented on x-pack/plugin/eql/qa/common/src/main/resources/test_queries_date_nanos.toml:
```diff
@@ -4,20 +4,20 @@
 name = "filterOnDateNanosWithMillis"
 query = '''
-    process where @timestamp == "1970-01-02T12:31:25.510Z"
+    process where @timestamp == "1974-03-02T19:53:16.510Z"
```
Why this change in test data? I got from you test changes that the data itself didn't actually change, I'm curious why have you changed these tests and, at the same time, `expected_event_ids` remained the same.

bpintea replied:
> I got from you test changes that the data itself didn't actually change

Yes, it did change: the nano tests used before a copy of `endgame-140.data` that was modified to match the few tests contained in `test_queries_date_nanos.toml`. With this PR `eql_date_nanos.data` has been removed and tests setup to use with the same data - `endgame-140.data` - [with different interpretation of the timestamp value](https://github.com/elastic/elasticsearch/pull/76953#discussion_r696092631). This allows to run the same tests - `test_queries.toml` on both millis and nanos. But this also requires the event nano tests - `test_queries_date_nanos.toml`  - changed in order to match the existing results (that are kept unchanged).

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java:
```diff
@@ -120,7 +126,21 @@ public boolean afterOrAt(Ordinal other) {
+    }
+
+    private static int timestampCompare(Object ts1, Object ts2) {
+        return ts1 instanceof Instant
+            ? (ts2 instanceof Instant ? ((Instant) ts1).compareTo((Instant) ts2) :
+                ((Instant) ts1).compareTo(Instant.ofEpochMilli(((Number) ts2).longValue())))
+            : (ts2 instanceof Instant ? Instant.ofEpochMilli(((Number) ts1).longValue()).compareTo((Instant) ts2) :
+                Long.compare(((Number) ts1).longValue(), ((Number) ts2).longValue()));
```
Isn't the ternary operator being a bit overused here? Wouldn't an `if/else` be more easy to follow visually?

Also, looking at the changes in this PR, having `instanceof Instant` logic separate as a utility method would make more sense to show the difference between a date-with-nanos and date-with-millis?

bpintea replied:
This has been refactored.

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/ImplicitTiebreakerTests.java:
```diff
@@ -65,7 +65,7 @@ public void query(QueryRequest r, ActionListener<SearchResponse> l) {
                 // except the first request, the rest should have the previous response's search_after _shard_doc value
                 assertArrayEquals("Elements at stage " + ordinal + " do not match",
-                    r.searchSource().searchAfter(), new Object[] { (long) previous, implicitTiebreakerValues.get(previous) });
+                    r.searchSource().searchAfter(), new Object[] { String.valueOf(previous), implicitTiebreakerValues.get(previous) });
```
Leftover from undo-ing the implicit tiebreaker changes?

bpintea replied:
This has be a string, since otherwise the value - before a Long/Instant, now a Timestamp - can't be used when processing downstream the search object.

astefan reviewed (approved):
LGTM in general. I've left some questions.

costin reviewed (changes requested):
The Instant/Long handling can be better encapsulated. I'm +1 in not having a wrapper class (see my other comment) around the Long/Instant however as it stands, the code that handles the differences is spread across multiple classes making it hard to follow.
Instead I propose to consolidate it in a single class (potentially in the ordinal package) and move there all the methods that perform comparison, difference, toString() etc...
That should simplify some things since constants like 1_000_000_000 are defined once and better documented while calling classes can simply call:
`timestamps.diff(ordinal.timestamp, newOrdinal.timestamp)`.

As a separate note, it's worth checking whether we can avoid some of the boxing/autoboxing happening by using `Object`. Previously `long` was used however due to `Instant` now `Object` is passed around as a signature which means anything that happens with `long` creates objects - so maybe adding a wrapping class (Timestamp) with different implementations, on for `long` and one for `Instant` might work out after all.

costin commented:
I'm concerned that the tests have to be changed to make this change work out - that should not be the case.

bpintea committed:
Address review comments

- wrap long and Instance timestamps into a Timestamp small hierarchy.
- small test fixes and commenting.

bpintea committed:
Merge branch 'master' into feat/eql_sequence_nanos

bpintea committed:
Address review comments

- cache MillisTimestamp conversion to Instant;
- merge and streamline some of the Timestamp methods.
- add comment on changed test

bpintea committed:
Merge branch 'master' into feat/eql_sequence_nanos

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:
```diff
@@ -175,14 +178,28 @@ private static void loadData(RestHighLevelClient client, String indexName, boole
+        String milliFraction = timestamp.substring(12);
+        // strip the fractions right away if not actually present
+        entry.put("@timestamp", milliFraction.equals("000000") ? millis : millis + "." + milliFraction);
+        entry.put("timestamp", ((long) object)/1_000_000L);
```
Just like for the existing millis tests, `timestamp` is populated but not used in the qa tests. Having it with a meaningful value allows however quick millis vs nanos comparison tests though (used mostly for troubleshooting, only changing the `timestamp_field`).

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:
```diff
@@ -86,33 +89,33 @@ public static void loadDatasetIntoEs(RestHighLevelClient client,
         // Date_Nanos index
         //
-        // The data for this index are identical to the endgame-140.data with only the values for @timestamp changed.
+        // The data for this index is loaded from the same endgame-140.data sample, only having the mapping for @timestamp changed.
```
Good point, thanks, added.

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:
```diff
@@ -44,8 +46,9 @@
- * - endgame-140 - for existing data
- * - extra       - additional data
+ * - endgame-140    - for existing data
+ * - eql_date_nanos - same as endgame-140, but with nano-precision timestamps
```
Applied.

bpintea commented on x-pack/plugin/eql/qa/common/src/main/resources/test_queries.toml:
```diff
@@ -861,7 +861,7 @@ expected_event_ids  = [67, 68, 69, 70,
 name = "sequencesOnDifferentEventTypes7"
 query = '''
-sequence with maxspan=500ms
+sequence with maxspan=50ms
```
I've added a comment to the test and [explained further](https://github.com/elastic/elasticsearch/pull/76953#discussion_r699568002) the change intention, to allow the same sequence queries be executed on both millis and nanos timestamps.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java:
```diff
@@ -42,14 +42,9 @@ public static Timestamp of(String milliseconds) {
 
-    // time delta in nanos from other to instance
+    // time delta in nanos between this and other instance
     public long diff(Timestamp other) {
```
This could be improved by `NANOS.between(this, other)` which behind the scenes does the math computation in this method.

costin replied:
Also delta, until or between sound better than `diff`.

bpintea replied:
Went for delta. (`until` suggests that `other` is expected to be past `this`, which is not and `between` might require a range argument)

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public int compareTo(Timestamp other) {
+        return other instanceof MillisTimestamp ? Long.compare(timestamp, ((MillisTimestamp) other).timestamp) : -other.compareTo(this);
```
to avoid the surprising `- ` you could just delegate to the parent - `super.compareTo(this)`.

bpintea replied:
Reverted to previous format.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public long diff(Timestamp other) {
+        return other instanceof MillisTimestamp ? (timestamp - ((MillisTimestamp) other).timestamp) * NANOS_PER_MILLI : -other.diff(this);
```
See above.

bpintea replied:
Ditto.

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceSpecTests.java:
```diff
@@ -85,8 +86,8 @@ public String hitName() {
         @Override
-        public Long extract(SearchHit hit) {
-            return (long) hit.docId();
+        public Object extract(SearchHit hit) {
```
The return time should be Timestamp not Object.

bpintea replied:
Fixed, thanks.

costin reviewed (approved):
Left some comments around cosmetics - LGTM.

bpintea committed:
Address review comments

- revert to using 'super' in MillisTimestamp;
- rename Timestamp#timestamp() to #instant().

bpintea committed:
Merge branch 'master' into feat/eql_sequence_nanos

bpintea committed:
Delegate nanos arithmetics

Delegate nanos arithmetics.

bpintea committed:
Remove no longer unused constant

Remove no longer unused constant.

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

To backport manually run `backport --upstream elastic/elasticsearch --pr 76953`
